Applications/Spire/Source/Ui/FocusObserver.cpp

```diff
@@ -0,0 +1,97 @@
+#include "Spire/Ui/FocusObserver.hpp"
+#include <QApplication>
+#include <QFocusEvent>
+
+using namespace boost::signals2;
+using namespace Spire;
+
+struct FocusObserver::FocusEventFilter : QObject {
+  FocusObserver* m_observer;
+  Qt::FocusReason m_focus_reason;
+
+  FocusEventFilter(FocusObserver& observer)
+      : QObject(),
+        m_observer(&observer),
+        m_focus_reason(Qt::MouseFocusReason) {
+    qApp->installEventFilter(this);
+    QObject::connect(qApp,
+      &QApplication::focusChanged, this, &FocusEventFilter::on_focus_changed);
+  }
+
+  bool eventFilter(QObject* watched, QEvent* event) override {
+    if(event->type() == QEvent::FocusIn && watched->isWidgetType() &&
+        m_observer->m_widget->isAncestorOf(static_cast<QWidget*>(watched))) {
+      m_focus_reason = static_cast<QFocusEvent*>(event)->reason();
+    }
+    return QObject::eventFilter(watched, event);
+  }
+
+  void on_focus_changed(QWidget* old, QWidget* now) {
+    static auto widget_focus_visible = std::pair<QWidget*, bool>();
+    static auto previous_widget_focus_visible = widget_focus_visible;
+    if(widget_focus_visible.first != now &&
+        previous_widget_focus_visible != widget_focus_visible) {
+      previous_widget_focus_visible = widget_focus_visible;
+    }
+    auto state = m_observer->m_state;
+    if(m_observer->m_widget == now) {
+      m_observer->m_state = State::FOCUS;
+    } else if(m_observer->m_widget->isAncestorOf(now)) {
+      m_observer->m_state = State::FOCUS_IN;
+    } else {
+      m_observer->m_state = State::NONE;
+    }
+    if(m_observer->m_state != State::NONE) {
```

Comment:
I believe the check should be ```state == State::FOCUS```. FOCUS_VISIBLE depends on FOCUS.

Reply:
I don't think so. For example, DurationBox contains multiple sub-components, when it gets focus by pressing Tab, the focus state is FOCUS_IN, then because the focus reason is Qt::TabFocusReason, the focus state should be updated FOCUS_VISIBLE. FOCUS_VISIBLE also implies being in the FOCUS_IN state

Reply:
Give me an example where ```m_observer->m_widget != now``` and ```m_observer->m_state == State::FOCUS_VISIBLE```

Reply:
The FocusObserver monitors the focus state of the DurationBox. The DurationBox contains three DecimalBoxs whose focus proxies are three TextBoxs. When pressing Tab, the first TextBox (The TextBox is the focus proxy of the first DruationBox) in the DurationBox gets focus, so now is TextBox and m_observer->m_widget is DurationBox (m_observer->m_widget != now). Because the DurationBox gets focus from keyboard, its focus state should be `State::FOCUS_VISIBLE`.

Reply:
It seems `focus_visible` is mutually exclusive to `focus_in` and `focus`. `foucs_in` and `focus` indicates a component or one of its descendants gets focus, whereas the `focus_visible` represents the component getting focus by keyboard. I think it would be easy to understand these states to me if the definition of `focus_in` is updated to “any of the box’s descendants has focus” by removing “the box has focus”, so `focus` doesn’t include `focus_in` again.

Reply:
The does not satisfy the definition of focus_visible given in the model. The model states that focus_visible is true when focus is true and either focus was received from a non-pointing device or focus was received from a component that itself had focus_visible.